---
output:
  pdf_document: default
  html_document: default
---

```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(ggthemes)
library(printr)
library(ggplot2)
library(cowplot)
library(reshape2)
library(knitr)
library(pwr)
library("ggpubr")
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_meta <- read.csv("~/master/wolf/dataflow/00-meta/wetzels2018_wu201fastq.csv")
colnames(df_meta)[1] <- "sra_accession"
df_meta$sra_accession <- as.character(df_meta$sra_accession)

df_classification  <- read.csv("~/master/wolf/dataflow/03-asv-taxonomy/wetzels2018_wu2018-97-rdp.csv")
df_classification[,1] <- NULL

df_97 <- read.delim("~/master/wolf/dataflow/03-asv-table-merge/wetzels2018_wu2018-97.txt", skip = 1, header = TRUE)
df_97$clustering <- 97

df <- df_97

colnames(df)[1] <- 'asv_id'

col_to_gather <- names(df)[!(startsWith(names(df), "SRR") |  startsWith(names(df), "W"))]
df <- melt(df, id = col_to_gather)
colnames(df)[3:4] <- c('sra_accession', 'count')

df$asv_id <- as.character(df$asv_id)
df_meta$type <- as.character(df_meta$type)
df_meta$study <- as.character(df_meta$study)
df_meta$region <- as.character(df_meta$region)
df_meta$sample <- as.character(df_meta$sample)

df_normalization <- df %>%
  filter(clustering == 97) %>%
  group_by(sra_accession) %>%
  mutate(total_16s_reads = sum(count)) %>%
  ungroup() %>%
  select(sra_accession, total_16s_reads) %>%
  ungroup() %>%
  distinct() 

df_classification <- df_classification %>%
  select(asv_id, phylum, family, genus)


df_complete <- inner_join(df, df_meta) %>%
  inner_join(df_normalization) %>%
  inner_join(df_classification)

df_normalization_plot <- inner_join(df_normalization, df_meta)
```

## summary

- The Wolf skin and fecal samples (Wolf Science Centre - WSC), as well as the Wolf and Dog fecal samples from Wu 2018, we're processed with dada2.
- Since the Wu 2018 data had longer reads after quality control, the data was trimmed to the same 275 basepairs as the WSC data. 
- All the data, after processing, was clustered together at 97% identity and taxonomically classified using RDB.

### number of sequences

- The first thing I noticed was that many of the samples from Wu 2018 have very few remaining reads after processing.
- The following plot is coloured by the study (our group or the Wu 2018 paper), with the area where the samples were collected on the x axis. 
- It is clear in Wu 2018 that there are far fewer sequences for the Henan area, so these will be removed. 
- Also, since there are no dogs in the WSC data I currently have and no Skin samples in the Wu 2018 data, I will be removing Dog and Skin samples for the sake of comparison.

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=7.5, fig.height=3}
plot <- ggplot(df_normalization_plot, aes(x=reorder(sra_accession, total_16s_reads))) +
    theme_few() +
    geom_point(aes(x=region, y = total_16s_reads, colour = study),
      stat = "identity", fill = "lightgrey", size = 1) +
    theme(strip.text = element_text(size = 12),
          axis.text.y = element_text(size = 12), 
          axis.title.y = element_text(size = 12)) +
    theme(axis.title.x=element_blank()) +
  ylab("Total Reads Afer DADA2")
plot
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_metrics <- df_complete %>% 
  filter(type == "Wolf") %>%
  filter(sample == "fecal") %>%
  
  filter(count > 0) %>%
  
  mutate(count_norm = (count / total_16s_reads)*100) %>%
  
  unite(classification, c("phylum", "family", "genus"), sep = ";", remove = FALSE)

df_metrics$classification <- gsub(";NA", "", df_metrics$classification)
```

\newpage

### phylum level grouping

- Looking at the percentage of the sample at a phylum level makes is very clear that the wolf communities from China and here are very distinct. This is especially true with the ratio of Bacteroidetes to Fermicutes (why is the Wolf feces here so Firmicute dominated?)


```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=7.5, fig.height=3}
df_metrics_phylum <- df_metrics %>%
  group_by(sra_accession, phylum) %>%
  mutate(phylum_percent = sum(count_norm)) %>%
  ungroup() %>%
  select(sra_accession, phylum, phylum_percent, study) %>%
  distinct()


ggplot() + geom_bar(
  aes(y = phylum_percent, x = sra_accession, fill = phylum),
  data = df_metrics_phylum,
  stat = "identity" ,
  position = "fill"
)  +labs(x="Category" , y= "Percentage (%)")+
  theme_minimal() +
  theme(axis.text.x = element_blank()) + 
     facet_wrap(.~study, scales="free")+
  ylab("Percentage of Sample") +
  xlab("Individual Sample")
```


### abundant sequences

- The next thing I noticed is that there are some very abundant OTUs in both datasets.
- Here are OTUs that represent more that 20% of their sample, coloured at the family level. It is clear even from looking at just a few of the abundant OTUs, that the Wolf Science scienter samples are very different from those in China.
- Also, in the inner mongolia data, there are 3 OTUs in 3 samples that make up almost 70% if the data. This is very strange. 

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=7.5, fig.height=3}
df_abundant_plot <- df_metrics %>%
  filter(count_norm > 20)


ggplot(data=df_abundant_plot, aes(x=reorder(sra_accession, count_norm), y=count_norm, fill=family)) +
  geom_bar(stat="identity")+
  scale_fill_brewer(palette="Paired")+
  theme_minimal() +
  theme(axis.text.x = element_blank())  + 
     facet_wrap(.~region, scales="free") +
  ylab("Percentage of Sample") +
  xlab("Individual Sample")
```