---
output:
  pdf_document: default
  html_document: default
---

```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(ggthemes)
library(printr)
library(ggplot2)
library(cowplot)
library(reshape2)
library(knitr)
library(pwr)
library("ggpubr")
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_meta <- read.csv("~/master/wolf/dataflow/00-meta/wetzels2018_wu2018_fastq.csv")
colnames(df_meta)[1] <- "sra_accession"
df_meta$sra_accession <- as.character(df_meta$sra_accession)
df_meta$type <- as.character(df_meta$type)
df_meta$study <- as.character(df_meta$study)
df_meta$region <- as.character(df_meta$region)
df_meta$sample <- as.character(df_meta$sample)
df_meta$pack <- as.character(df_meta$pack)
df_meta$sibling <- as.character(df_meta$sibling)
df_meta$replicate <- as.character(df_meta$replicate)
df_meta$sample_id <- as.character(df_meta$sample_id)

df_classification  <- read.csv("~/master/wolf/dataflow/03-asv-taxonomy/wetzels2018_wu2018-97-rdp.csv")
df_classification[,1] <- NULL

df <- read.delim("~/master/wolf/dataflow/03-asv-table-merge/wetzels2018_wu2018-97.txt", skip = 1, header = TRUE)
df$clustering <- 97

colnames(df)[1] <- 'asv_id'

col_to_gather <- names(df)[!(startsWith(names(df), "SRR") |  startsWith(names(df), "W"))]
df <- melt(df, id = col_to_gather)
colnames(df)[3:4] <- c('sra_accession', 'count')

df$asv_id <- as.character(df$asv_id)


df_normalization <- df %>%
  filter(clustering == 97) %>%
  group_by(sra_accession) %>%
  mutate(total_16s_reads = sum(count)) %>%
  ungroup() %>%
  select(sra_accession, total_16s_reads) %>%
  ungroup() %>%
  distinct() 

df_classification <- df_classification %>%
  select(asv_id, phylum, family, genus)


df_complete <- inner_join(df, df_meta) %>%
  inner_join(df_normalization) %>%
  inner_join(df_classification)

df_normalization_plot <- inner_join(df_normalization, df_meta)
```

## summary

- The Wolf skin and fecal samples (Wolf Science Centre - WSC), as well as the Wolf and Dog fecal samples from Wu 2018, we're processed with dada2.
- Since the Wu 2018 data had longer reads after quality control, the data was trimmed to the same 275 basepairs as the WSC data. 
- All the data, after processing, was clustered together at 97% identity and taxonomically classified using RDB.

### number of sequences

- The first thing I noticed was that many of the samples from Wu 2018 have very few remaining reads after processing.
- The following plot is coloured by the study (our group or the Wu 2018 paper), with the area where the samples were collected on the x axis. 
- It is clear in Wu 2018 that there are far fewer sequences for the Henan area, so these will be removed. 
- Also, since there are no dogs in the WSC data I currently have and no Skin samples in the Wu 2018 data, I will be removing Dog and Skin samples for the sake of comparison.

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=7.5, fig.height=3}
plot <- ggplot(df_normalization_plot, aes(x=reorder(sra_accession, total_16s_reads))) +
    theme_few() +
    geom_point(aes(x=region, y = total_16s_reads, colour = study),
      stat = "identity", fill = "lightgrey", size = 1) +
    theme(strip.text = element_text(size = 12),
          axis.text.y = element_text(size = 12), 
          axis.title.y = element_text(size = 12)) +
    theme(axis.title.x=element_blank()) +
  ylab("Total Reads Afer DADA2")
plot
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_metrics <- df_complete %>% 

  filter(region != "Henan") %>%
  
  mutate(count_norm = (count / total_16s_reads)*100) %>%
  
  group_by(sample_id, asv_id) %>%
  
  mutate(count_norm = median(count_norm)) %>%
  
  ungroup() %>%
  
  unite(classification, c("phylum", "family", "genus"), sep = ";", remove = FALSE) %>%
  
  select(-sra_accession, -replicate, -total_16s_reads, -count) %>%
  distinct()

df_metrics_wsc <- df_metrics %>%
  filter(study == "wetzels2018") %>%
  filter(sample == "fecal") %>%
  filter(pack == "W_pack_B" | pack == "W_pack_D" | pack == "W_pack_E")

df_metrics_wsc2 <- df_metrics %>%
  filter(study == "wetzels2018") %>%
  filter(sample == "skin") %>%
  filter(pack == "W_pack_B" | pack == "W_pack_D" | pack == "W_pack_E")

df_metrics <- df_metrics %>% 
  #filter(type == "Wolf") %>%
  filter(sample == "fecal")

  #df_metrics$classification <- gsub(";NA", "", df_metrics$classification)
```

\newpage

### phylum level grouping

- Looking at the percentage of the sample at a phylum level makes is very clear that the wolf communities from China and here are very distinct.  
- There is one sample from the WSC (W4f) that looks a bit more like the Wu data, based on the amount of bacteroidetes, fusobacteria and actinobacteria predominatly.
- Further, dogs and wolfs from china seems to be more similar as wolfs from the WSC. 

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=10, fig.height=7}
df_metrics_phylum <- df_metrics %>%
  filter(count_norm > 0.1) %>%
  group_by(sample_id, phylum) %>%
  mutate(phylum_percent = sum(count_norm)) %>%
  ungroup() %>%
  select(sample_id, phylum, phylum_percent, study, region, type) %>%
  distinct() 

ggplot() + geom_bar(
  aes(y = phylum_percent, x = sample_id, fill = phylum),
  data = df_metrics_phylum,
  stat = "identity" ,
  position = "fill"
)  +labs(x="Category" , y= "Percentage (%)")+
  theme_minimal() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 8)) + 
     facet_wrap(.~study, scales="free")+
  ylab("Percentage of Sample") +
  xlab("Individual Sample")
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_phlycounts_counts <- df_metrics %>%
  select(asv_id, sample_id, count_norm) %>%
  spread(sample_id, count_norm)

df_phlycounts_counts <- df_phlycounts_counts[order(df_phlycounts_counts$asv_id),] 

otumat <- as.matrix(df_phlycounts_counts[,2:29])
rownames(otumat) <- as.data.frame(df_phlycounts_counts)[,1]

df_phlycounts_tax <- df_metrics %>%
  select(asv_id, phylum, family, genus) %>%
  unique()

df_phlycounts_tax$phylum <- as.character(df_phlycounts_tax$phylum)
df_phlycounts_tax$family <- as.character(df_phlycounts_tax$family)
df_phlycounts_tax$genus <- as.character(df_phlycounts_tax$genus)

df_phlycounts_tax[is.na(df_phlycounts_tax)] <- "NotAssigned"

df_phlycounts_tax <- df_phlycounts_tax[order(df_phlycounts_tax$asv_id),] 

taxmat <- as.matrix(df_phlycounts_tax[,2:4])
rownames(taxmat) <- as.data.frame(df_phlycounts_tax)[,1]

df_phylocounts_meta <- df_metrics %>% 
  select(sample_id, study, region, type) %>%
  unique() %>%
  rowwise() %>%
  ungroup()

df_phylocounts_meta <- as.data.frame(df_phylocounts_meta)

row.names(df_phylocounts_meta) <- df_phylocounts_meta[,1]
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_phlycounts_counts_wsc <- df_metrics_wsc %>%
  select(asv_id, sample_id, count_norm) %>%
  spread(sample_id, count_norm)

df_phlycounts_counts_wsc <- df_phlycounts_counts_wsc[order(df_phlycounts_counts_wsc$asv_id),] 

otumat_wsc <- as.matrix(df_phlycounts_counts_wsc[,2:10])
rownames(otumat_wsc) <- as.data.frame(df_phlycounts_counts_wsc)[,1]

df_phlycounts_tax_wsc <- df_metrics_wsc %>%
  select(asv_id, phylum, family, genus) %>%
  unique()

df_phlycounts_tax_wsc$phylum <- as.character(df_phlycounts_tax_wsc$phylum)
df_phlycounts_tax_wsc$family <- as.character(df_phlycounts_tax_wsc$family)
df_phlycounts_tax_wsc$genus <- as.character(df_phlycounts_tax_wsc$genus)

df_phlycounts_tax_wsc[is.na(df_phlycounts_tax_wsc)] <- "NotAssigned"

df_phlycounts_tax_wsc <- df_phlycounts_tax_wsc[order(df_phlycounts_tax_wsc$asv_id),] 

taxmat_wsc <- as.matrix(df_phlycounts_tax_wsc[,2:4])
rownames(taxmat_wsc) <- as.data.frame(df_phlycounts_tax_wsc)[,1]

df_phylocounts_meta_wsc <- df_metrics_wsc %>% 
  select(sample_id, pack, sibling) %>%
  unique() %>%
  rowwise() %>%
  ungroup()

df_phylocounts_meta_wsc <- as.data.frame(df_phylocounts_meta_wsc)

row.names(df_phylocounts_meta_wsc) <- df_phylocounts_meta_wsc[,1]
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_phlycounts_counts_wsc2 <- df_metrics_wsc2 %>%
  select(asv_id, sample_id, count_norm) %>%
  spread(sample_id, count_norm)

df_phlycounts_counts_wsc2 <- df_phlycounts_counts_wsc2[order(df_phlycounts_counts_wsc2$asv_id),] 

otumat_wsc2 <- as.matrix(df_phlycounts_counts_wsc2[,2:10])
rownames(otumat_wsc2) <- as.data.frame(df_phlycounts_counts_wsc2)[,1]

df_phlycounts_tax_wsc2 <- df_metrics_wsc2 %>%
  select(asv_id, phylum, family, genus) %>%
  unique()

df_phlycounts_tax_wsc2$phylum <- as.character(df_phlycounts_tax_wsc2$phylum)
df_phlycounts_tax_wsc2$family <- as.character(df_phlycounts_tax_wsc2$family)
df_phlycounts_tax_wsc2$genus <- as.character(df_phlycounts_tax_wsc2$genus)

df_phlycounts_tax_wsc2[is.na(df_phlycounts_tax_wsc2)] <- "NotAssigned"

df_phlycounts_tax_wsc2 <- df_phlycounts_tax_wsc2[order(df_phlycounts_tax_wsc2$asv_id),] 

taxmat_wsc2 <- as.matrix(df_phlycounts_tax_wsc2[,2:4])
rownames(taxmat_wsc2) <- as.data.frame(df_phlycounts_tax_wsc2)[,1]

df_phylocounts_meta_wsc2 <- df_metrics_wsc2 %>% 
  select(sample_id, pack, sibling) %>%
  unique() %>%
  rowwise() %>%
  ungroup()

df_phylocounts_meta_wsc2 <- as.data.frame(df_phylocounts_meta_wsc2)

row.names(df_phylocounts_meta_wsc2) <- df_phylocounts_meta_wsc2[,1]
```


```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=20, fig.height=8}
library("phyloseq")
library("ape")
library(plyr)
OTU = otu_table(otumat, taxa_are_rows = TRUE)
TAX = tax_table(taxmat)
physeq = phyloseq(OTU, TAX)
random_tree = rtree(ntaxa(physeq), rooted=TRUE, tip.label=taxa_names(physeq))
samplesdata <- sample_data(df_phylocounts_meta)
physeq = phyloseq(OTU, TAX, samplesdata, random_tree)
#plot_bar(physeq, fill = "phylum")
#physeq
```

This is quite obvious is I make a network, where W4f calls between the WSC samples and the China samples.  

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=5, fig.height=8}
net <- make_network(physeq, "samples", max.dist=0.8)
plot_network(net, physeq , color="sample_id", shape="study", line_weight=0.3, label=NULL)
```

- We can also build all the different types of ordinations. Here, I will make the shape the region, as the two different regions in China seems to be more similar, and colour by the study. In the second version of the plot, I will colour by animals, as the dogs to stand out in a few ordinations. 

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=10, fig.height=10}
phylum.sum = tapply(taxa_sums(physeq), tax_table(physeq)[, "phylum"], sum, na.rm=TRUE)
top5phyla = names(sort(phylum.sum, TRUE))[1:11]
physeq1 = prune_taxa((tax_table(physeq)[, "phylum"] %in% top5phyla), physeq)

dist = "bray"
ord_meths = c("DCA", "CCA", "RDA", "DPCoA", "NMDS", "MDS", "PCoA")
plist = llply(as.list(ord_meths), function(i, physeq, dist){
        ordi = ordinate(physeq, method=i, distance=dist)
        plot_ordination(physeq, ordi, "samples", color="study")
}, physeq1, dist)
names(plist) <- ord_meths

pdataframe = ldply(plist, function(x){
    df = x$data[, 1:2]
    colnames(df) = c("Axis_1", "Axis_2")
    return(cbind(df, x$data))
})
names(pdataframe)[1] = "method"

p = ggplot(pdataframe, aes(Axis_1, Axis_2, color=study, shape=region, fill=study))
p = p + geom_point(size=4) + geom_polygon()
p = p + facet_wrap(~method, scales="free")
#p = p + scale_fill_brewer(type="qual", palette="Set1")
#p = p + scale_colour_brewer(type="qual", palette="Set1")
p



plist = llply(as.list(ord_meths), function(i, physeq, dist){
        ordi = ordinate(physeq, method=i, distance=dist)
        plot_ordination(physeq, ordi, "samples", color="type")
}, physeq1, dist)
names(plist) <- ord_meths

pdataframe2 = ldply(plist, function(x){
    df = x$data[, 1:2]
    colnames(df) = c("Axis_1", "Axis_2")
    return(cbind(df, x$data))
})
names(pdataframe2)[1] = "method"

p2 = ggplot(pdataframe, aes(Axis_1, Axis_2, color=type, shape=region, fill=type))
p2 = p2 + geom_point(size=4) #+ geom_polygon()
p2 = p2 + facet_wrap(~method, scales="free")
#p = p + scale_fill_brewer(type="qual", palette="Set1")
#p = p + scale_colour_brewer(type="qual", palette="Set1")
p2
```

- I think the conclusion here is that the samples from China and the WSC are very different, with a single wolf from the WSC that appears to cluster quite a bit closer to the Chinese dogs and wolfs. 

### only looking at WSC feces data

- I will generate the ordinations again, with the first plots coloured by pack and the second coloured by sibling groups.
- Maybe a few of the packs a clustering a bit? For example, in the NMDS, 3 packs may cluster together - but nothing clear. 

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE}
OTU_wsc = otu_table(otumat_wsc, taxa_are_rows = TRUE)
TAX_wsc = tax_table(taxmat_wsc)
physeq_wsc = phyloseq(OTU_wsc, TAX_wsc)
random_tree_wsc = rtree(ntaxa(physeq_wsc), rooted=TRUE, tip.label=taxa_names(physeq_wsc))
samplesdata_wsc <- sample_data(df_phylocounts_meta_wsc)
physeq_wsc = phyloseq(OTU_wsc, TAX_wsc, samplesdata_wsc, random_tree_wsc)
#plot_bar(physeq, fill = "phylum")
#physeq_wsc
```




```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=10, fig.height=10}
phylum.sum = tapply(taxa_sums(physeq_wsc), tax_table(physeq_wsc)[, "phylum"], sum, na.rm=TRUE)
top5phyla = names(sort(phylum.sum, TRUE))[1:11]
physeq1_wsc = prune_taxa((tax_table(physeq_wsc)[, "phylum"] %in% top5phyla), physeq_wsc)

dist = "bray"
ord_meths = c("DCA", "CCA", "RDA", "DPCoA", "NMDS", "MDS", "PCoA")
plist = llply(as.list(ord_meths), function(i, physeq_wsc, dist){
        ordi = ordinate(physeq_wsc, method=i, distance=dist)
        plot_ordination(physeq_wsc, ordi, "samples", color="pack")
}, physeq1_wsc, dist)
names(plist) <- ord_meths

pdataframe = ldply(plist, function(x){
    df = x$data[, 1:2]
    colnames(df) = c("Axis_1", "Axis_2")
    return(cbind(df, x$data))
})
names(pdataframe)[1] = "method"

p = ggplot(pdataframe, aes(Axis_1, Axis_2, color=pack, fill=pack))
p = p + geom_point(size=4) + geom_polygon()
p = p + facet_wrap(~method, scales="free")
#p = p + scale_fill_brewer(type="qual", palette="Set1")
#p = p + scale_colour_brewer(type="qual", palette="Set1")
p



plist = llply(as.list(ord_meths), function(i, physeq, dist){
        ordi = ordinate(physeq, method=i, distance=dist)
        plot_ordination(physeq, ordi, "samples", color="sibling")
}, physeq1, dist)
names(plist) <- ord_meths

pdataframe2 = ldply(plist, function(x){
    df = x$data[, 1:2]
    colnames(df) = c("Axis_1", "Axis_2")
    return(cbind(df, x$data))
})
names(pdataframe2)[1] = "method"

p2 = ggplot(pdataframe, aes(Axis_1, Axis_2, color=sibling, fill=sibling))
p2 = p2 + geom_point(size=4) + geom_polygon()
p2 = p2 + facet_wrap(~method, scales="free")
#p = p + scale_fill_brewer(type="qual", palette="Set1")
#p = p + scale_colour_brewer(type="qual", palette="Set1")
p2
```


### only looking at WSC skin data

- I will generate the ordinations now for the skin, with the first plots coloured again by sample and the second coloured by sibling groups.

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE}
OTU_wsc2 = otu_table(otumat_wsc2, taxa_are_rows = TRUE)
TAX_wsc2 = tax_table(taxmat_wsc2)
physeq_wsc2 = phyloseq(OTU_wsc2, TAX_wsc2)
random_tree_wsc2 = rtree(ntaxa(physeq_wsc2), rooted=TRUE, tip.label=taxa_names(physeq_wsc2))
samplesdata_wsc2 <- sample_data(df_phylocounts_meta_wsc2)
physeq_wsc2 = phyloseq(OTU_wsc2, TAX_wsc2, samplesdata_wsc2, random_tree_wsc2)
#plot_bar(physeq, fill = "phylum")
#physeq_wsc2
```




```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=10, fig.height=10}
phylum.sum = tapply(taxa_sums(physeq_wsc2), tax_table(physeq_wsc2)[, "phylum"], sum, na.rm=TRUE)
top5phyla = names(sort(phylum.sum, TRUE))[1:11]
physeq1_wsc2 = prune_taxa((tax_table(physeq_wsc2)[, "phylum"] %in% top5phyla), physeq_wsc2)

dist = "bray"
ord_meths = c("DCA", "CCA", "RDA", "DPCoA", "NMDS", "MDS", "PCoA")
plist = llply(as.list(ord_meths), function(i, physeq_wsc2, dist){
        ordi = ordinate(physeq_wsc2, method=i, distance=dist)
        plot_ordination(physeq_wsc2, ordi, "samples", color="pack")
}, physeq1_wsc2, dist)
names(plist) <- ord_meths

pdataframe = ldply(plist, function(x){
    df = x$data[, 1:2]
    colnames(df) = c("Axis_1", "Axis_2")
    return(cbind(df, x$data))
})
names(pdataframe)[1] = "method"

p = ggplot(pdataframe, aes(Axis_1, Axis_2, color=pack, fill=pack))
p = p + geom_point(size=4) + geom_polygon()
p = p + facet_wrap(~method, scales="free")
#p = p + scale_fill_brewer(type="qual", palette="Set1")
#p = p + scale_colour_brewer(type="qual", palette="Set1")
p



plist = llply(as.list(ord_meths), function(i, physeq, dist){
        ordi = ordinate(physeq, method=i, distance=dist)
        plot_ordination(physeq, ordi, "samples", color="sibling")
}, physeq1, dist)
names(plist) <- ord_meths

pdataframe2 = ldply(plist, function(x){
    df = x$data[, 1:2]
    colnames(df) = c("Axis_1", "Axis_2")
    return(cbind(df, x$data))
})
names(pdataframe2)[1] = "method"

p2 = ggplot(pdataframe, aes(Axis_1, Axis_2, color=sibling, fill=sibling))
p2 = p2 + geom_point(size=4) + geom_polygon()
p2 = p2 + facet_wrap(~method, scales="free")
#p = p + scale_fill_brewer(type="qual", palette="Set1")
#p = p + scale_colour_brewer(type="qual", palette="Set1")
p2
```

