---
output:
  pdf_document: default
  word_document: default
  html_document: default
---
```{r, echo=FALSE, warning=FALSE, message=FALSE}
library(tidyverse)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
# colnames(df_num_genes)[1] <- "file"
# 
# resistances <- read.delim("~/master/rumen/dataflow/02-blast/rumen_genomes_card.txt", header=FALSE) 
# aro_categories <- read.delim("~/master/databases/aro_categories_index.csv")
# 
# colnames(resistances) <- c("qseqid", "sseqid", "pident", "sstart", "send", "qstart", "qend", "evalue", "bitscore", "score", "qlen", "length", "sseq")
# 
# resistances <- resistances %>%
#   select(-sseq) %>%
#   separate(sseqid, into = c("rm", "Protein.Accession", "ARO.Accession", "gene"), sep = "\\|", remove = FALSE) 
# 
# 
# resistances$gene <- gsub("_", " ", resistances$gene)
# 
# 
# resistances <- resistances %>%
#     separate(gene, into = c("gene", "organism"), sep = "\\[", remove = FALSE) 
# 
# resistances$organism <- gsub("\\]", " ", resistances$organism)
# 
# resistances <- resistances %>%
#   inner_join(aro_categories) %>%
#   rowwise() %>%
#   mutate(file = stri_reverse(stri_split_fixed(stri_reverse(qseqid),"_",n = 3)[[1]][3])) %>%
#   ungroup() %>%
#   filter(pident > 50) %>%
#   filter(length > 50) %>%
#   #filter(bitscore >= 50) %>%
#   filter(Resistance.Mechanism == "antibiotic inactivation") %>%
#   group_by(gene) %>%
#   mutate(gene_occurences = length(file)) %>%
#   ungroup()
# 
# resistances_summary <- resistances %>%
#   filter(pident == 100) %>%
#   group_by(file) %>%
#   mutate(nfiles = length(file)) %>%
#   ungroup() %>%
#   filter(nfiles > 1)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
# df_blast_hits <- read.csv("~/master/rumen/dataflow/00-meta/resistance_island_blast.csv")
# 
# df_blast_hits_CP033865.1 <- df_blast_hits %>%
#   filter(accession == "CP033865.1") %>%
#   rowwise() %>%
#   mutate(duplicate = ifelse(start < 400000, 1, 2)) %>%
#   ungroup()
# 
# df_blast_hits$duplicate <- 1
# 
# df_blast_hits <- df_blast_hits %>%
#   filter(accession != "CP033865.1") 
# 
# 
# df_blast_hits <- bind_rows(df_blast_hits, df_blast_hits_CP033865.1)
# 
# df_blast_hits <- df_blast_hits %>%
#   select(-query, -pident) %>%
#   group_by(accession, duplicate) %>%
#   mutate(lowest_start = min(start)) %>%
#   mutate(highest_start = max(start)) %>%
#   mutate(lowest_stop = min(stop)) %>%
#   mutate(highest_stop = max(stop)) %>%
#   ungroup() %>%
#   select(-start, -stop) %>%
#   gather(name, value, -accession, -duplicate) %>%
#   group_by(accession, duplicate) %>%
#   mutate(start = min(value)) %>%
#   mutate(stop = max(value)) %>%
#   ungroup() %>%
#   select(-name, -value) %>%
#   mutate(length = stop - start) %>%
#   distinct() %>%
#   unite(ID, c("accession", "duplicate"), sep = "_", remove = FALSE)
#   
# df_blast_hits$direction <- "+"
# 
# colnames(df_blast_hits)[2] <- "contig"
# colnames(df_blast_hits)[5] <- "end"
# 
# write.csv(df_blast_hits, "~/master/rumen/dataflow/00-meta/resistance_blast_hit_cotigs.csv")
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
# island_mapping_all <- read.delim("~/master/rumen/dataflow/02-blast/resistance_island_blast_hits_concatenated_extractedCONTIGs_3rumen_all_genomes.txt", header=FALSE)
# colnames(island_mapping_all) <- c("qseqid", "sseqid", "pident", "sstart", "send", "qstart", "qend", "evalue", "bitscore", "score", "qlen", "length", "sseq")
# 
# 
# island_mapping_all <- island_mapping_all %>%
#   filter(sseqid != "4309680-submission.assembly_52") %>%
#   filter(sseqid != "4309689-submission.assembly_79") %>%
#   filter(sseqid != "RUG782_1") %>%
#   filter(pident > 80) %>%
#   group_by(qseqid, sseqid) %>%
#   mutate(sum_length = sum(length)) %>%
#   ungroup() %>%
#   filter(sum_length > 4000)
# 
# island_mapping_all_contigs <- island_mapping_all %>%
#   select(sseqid, sstart, send) %>%
#   group_by(sseqid) %>%
#   mutate(lowest_start = min(sstart)) %>%
#   mutate(highest_start = max(sstart)) %>%
#   mutate(lowest_stop = min(send)) %>%
#   mutate(highest_stop = max(send)) %>%
#   ungroup() %>%
#   select(-sstart, -send) %>%
#   gather(name, value, -sseqid) %>%
#   group_by(sseqid) %>%
#   mutate(start = min(value)) %>%
#   mutate(stop = max(value)) %>%
#   ungroup() %>%
#   select(-name, -value) %>%
#   mutate(length = stop - start) %>%
#   distinct()
# 
# island_mapping_all_contigs$direction <- "+"
# 
# colnames(island_mapping_all_contigs)[1] <- "contig"
# colnames(island_mapping_all_contigs)[3] <- "end"
# 
# island_mapping_all_contigs$ID <- island_mapping_all_contigs$contig
# 
# write.csv(island_mapping_all_contigs, "~/master/rumen/dataflow/00-meta/resistance_blast_hit_cotigs_all_rumen.csv")

```

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=7}
set.seed(1)
island_mapping_ava <- read.delim("~/master/rumen/dataflow/02-blast/resistance_island_mapping_allvall.txt", header=FALSE)
colnames(island_mapping_ava) <- c("qseqid", "sseqid", "pident", "sstart", "send", "qstart", "qend", "evalue", "bitscore", "score", "qlen", "length", "sseq")

island_mapping_ava$qseqid <- as.character(island_mapping_ava$qseqid)
island_mapping_ava$sseqid <- as.character(island_mapping_ava$sseqid)

island_mapping_ava <- island_mapping_ava %>%
  mutate(same = if_else(qseqid == sseqid, "yes", "no")) %>%
  filter(same != "yes") %>%
  filter(qseqid != "4309689-submission.assembly_79_resistance_genes") %>%
  filter(qseqid != "4309680-submission.assembly_52_resistance_genes") %>%
  filter(qseqid != "RUG782_1_resistance_genes") %>%
  mutate(same_seq = if_else(pident == 100 & qlen == length, "yes", "no")) %>%
  filter(same_seq == "yes") %>%
  group_by(qseqid) %>%
  mutate(group_num = length(unique(sseqid)) + 1) %>%
  ungroup() %>%
  select(qseqid, sseqid, group_num) %>%
  group_by(qseqid) %>%
  summarize(group_members = list(sort(c(sseqid, unique(qseqid))))) %>%
  ungroup() %>%
  select(-qseqid) %>%
  distinct()

list_groups <- unique(island_mapping_ava$group_members)
rep_seqs <- c()
num_in_group <- c()

for (i in 1:length(list_groups)){
  group_list <- list_groups[[i]]
  rnum <- sample(1:length(group_list), 1)
  rep_seqs <- c(rep_seqs, group_list[[rnum]])
  num_in_group <- c(num_in_group, length(group_list))
  
}

df_rep_seqs <- as.data.frame(cbind(rep_seqs, num_in_group))
colnames(df_rep_seqs)[1] <- "qseqid"

rem_seqs <- unlist(list_groups)[ !(unlist(list_groups) %in% df_rep_seqs$qseqid)]

island_mapping <- read.delim("~/master/rumen/dataflow/02-blast/resistance_island_mapping2.txt", header=FALSE)
colnames(island_mapping) <- c("qseqid", "sseqid", "pident", "sstart", "send", "qstart", "qend", "evalue", "bitscore", "score", "qlen", "length", "sseq")

df_agsinst_RUG782 <- island_mapping %>%
  select(qseqid, sseqid, pident, sstart, send, qstart, qend, bitscore, length) %>%
  filter(sseqid == "RUG782_1_resistance_genes") %>%
  mutate(nstart = if_else(sstart < send, sstart, send)) %>%
  mutate(nsend = if_else(send > sstart, send, sstart)) %>%
  ungroup() %>%
  filter(qseqid != "RUG782_1_resistance_genes") %>%
  group_by(qseqid) %>%
  mutate(bitscore_sum = sum(bitscore)) %>%
  ungroup() %>%
  filter(sstart > 110000) %>%
  group_by(qseqid, sseqid) %>%
  mutate(length_sum = sum(length)) %>%
  ungroup() %>%
  filter(length_sum > 4000) %>%
  filter(pident > 80)

# 4000 bp only cuts off a single clone, so I could say top blast hits with that minimum length? could also have a PI of 80%. 

df_agsinst_RUG782_rumen <- df_agsinst_RUG782 %>%
  filter(qseqid == "4309689-submission.assembly_79_resistance_genes" | qseqid == "4309680-submission.assembly_52_resistance_genes" | qseqid == "RUG782_1_resistance_genes" | qseqid == "4309680-submission.assembly_14" | qseqid == "4309689-submission.assembly_8" |qseqid == "RUG457_28" |qseqid == "RUG457_82" |qseqid == "RUG457_79" |qseqid == "RMG_164_291" | qseqid == "RUG459_26" | qseqid == "RMG_238_348" | qseqid == "RMG_238_348" | qseqid == "RUG459_38" | qseqid == "RUG459_38" | qseqid == "RMG_164_121")
  

df_agsinst_RUG782 <- df_agsinst_RUG782 %>%
  full_join(df_rep_seqs) %>%
  bind_rows(df_agsinst_RUG782_rumen) %>%
  mutate(remove = qseqid %in% rem_seqs) %>%
  filter(remove != TRUE) 

df_agsinst_RUG782$newqseqid <- NA
df_agsinst_RUG782$num_in_group <- as.character(df_agsinst_RUG782$num_in_group)
df_agsinst_RUG782 <- as.data.frame(df_agsinst_RUG782)

for (j in 1:nrow(df_agsinst_RUG782)) {
  test <- df_agsinst_RUG782[j, "num_in_group"]
  if(is.na(test)) {
    df_agsinst_RUG782[j, "newqseqid"] <- df_agsinst_RUG782[j, "qseqid"]} else {
      df_agsinst_RUG782[j, "newqseqid"] <- paste(df_agsinst_RUG782[j, "qseqid"], " (", df_agsinst_RUG782[j, "num_in_group"], ")", sep = "")
    }
  }



df_agsinst_RUG782 <- df_agsinst_RUG782 %>%
  arrange(desc(bitscore_sum))
  

df_agsinst_RUG782$newqseqid <- gsub("-submission.assembly", "", df_agsinst_RUG782$newqseqid)
df_agsinst_RUG782$newqseqid <- gsub("_resistance_genes", "", df_agsinst_RUG782$newqseqid)

df_agsinst_RUG782$newqseqid <- factor(df_agsinst_RUG782$newqseqid, levels = unique(as.character(df_agsinst_RUG782$newqseqid)))

df_seq_species_type <- read.csv("~/master/rumen/dataflow/00-meta/resistance_island_seq_origins.csv")
df_seq_species_type$accession <- as.character(df_seq_species_type$accession)

df_agsinst_RUG782 <- df_agsinst_RUG782 %>%
  filter(qseqid != "AJ557257.1_1") %>%
  separate(qseqid, into = c("accession", "rm"), remove = FALSE, sep = "_") %>%
  inner_join(df_seq_species_type, by = c("accession")) %>%
  unite(newqseqid2, c("species", "type", "newqseqid"), sep = " | ", remove = FALSE)

df_agsinst_RUG782$newqseqid2 <- factor(df_agsinst_RUG782$newqseqid2, levels = unique(as.character(df_agsinst_RUG782$newqseqid2)))

library(ggplot2)
library(ggalt)
theme_set(theme_classic())

# gg <- ggplot(df_agsinst_RUG782, aes(x=nstart, xend=nsend, y=newqseqid, group=newqseqid)) + 
#         geom_dumbbell(size=2, aes(color = pident)) + 
#         #scale_x_continuous(label=percent) + 
#         labs(x=NULL, 
#              y=NULL, 
#              title="Dumbbell Chart") +
#         theme(plot.title = element_text(hjust=0.5, face="bold"),
#               plot.background=element_rect(fill="#f7f7f7"),
#               panel.background=element_rect(fill="#f7f7f7"),
#               panel.grid.minor=element_blank(),
#               panel.grid.major.y=element_blank(),
#               panel.grid.major.x=element_line(),
#               axis.ticks=element_blank(),
#               legend.position="top",
#               panel.border=element_blank()) 




gg <- ggplot(df_agsinst_RUG782, 
       aes(y = newqseqid2,
           x = nstart,
           xend = nsend)) +  
  geom_dumbbell(size = 1.5,
                size_x = 1.5, 
                size_xend = 1.5,
                colour_x = "red", 
                colour_xend = "red", aes(colour = pident, colour_x = pident, colour_xend = pident)) +
  theme_minimal() + 
  labs(x = "Mapping to RUG782_1",
       y = "") +
    theme(text = element_text(size=8)) 
  

plot(gg)

df_unique_contigs <- df_agsinst_RUG782 %>%
  select(qseqid) %>%
  distinct()

#write.csv(df_unique_contigs, "~/master/rumen/dataflow/00-meta/resistance_blast_hit_cotigs_unique.csv")
```


```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=2.5, fig.width=5}
df_agsinst_RUG782_frag1 <- df_agsinst_RUG782 %>%
  filter(nstart == 132252 & nsend == 133739)

df_agsinst_RUG782_frag1$frag <- 1

df_agsinst_RUG782_frag2 <- df_agsinst_RUG782 %>%
  filter(nstart == 137348 & nsend == 140714)

df_agsinst_RUG782_frag2$frag <- 2

df_agsinst_RUG782_frag3 <- df_agsinst_RUG782 %>%
  filter(nstart == 141968 & nsend == 143706 )

df_agsinst_RUG782_frag3$frag <- 3


df_agsinst_RUG782_frags <- bind_rows(df_agsinst_RUG782_frag1, df_agsinst_RUG782_frag2, df_agsinst_RUG782_frag3) %>%
  group_by(frag) %>%
  mutate(avg_pident = mean(pident)) %>%
  ungroup()
```

\newpage

```{r, echo=FALSE, warning=FALSE, message=FALSE}
library(printr)

df_group_list <- list()
j <- 1

for (i in 1:length(list_groups)){
  
  df_group <- as.data.frame(list_groups[[i]])
  colnames(df_group) <- "qseqid"
  df_group$group <- i
  df_group_list[[j]] <- df_group
  j <- j + 1
  
}

df_group <- bind_rows(df_group_list)

df_seq_origin <- read.csv("~/master/rumen/dataflow/00-meta/resistance_island_sequences.csv")

df_group_seq_origin <- island_mapping %>%
  select(qseqid) %>%
  distinct() %>%
  separate(qseqid, c("accession", "contig_num"), sep = "_", remove = FALSE) %>%
  filter(qseqid != "4309689-submission.assembly_79_resistance_genes") %>%
  filter(qseqid != "4309680-submission.assembly_52_resistance_genes") %>%
  filter(qseqid != "RUG782_1_resistance_genes") %>%
  full_join(df_seq_origin) %>%
  full_join(df_group) %>%
  arrange(group)

#df_group_seq_origin %>%
#  select(qseqid, group, seq)


```

\newpage

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.5, fig.width=10}
library(genoPlotR)

xlims <- list(c(-Inf, Inf), c(-Inf, Inf), c(-Inf, Inf),  c(45000, 0))

df_seg_1 <- dna_seg(read.csv("~/master/rumen/dataflow/00-meta/RUG782.csv"))
df_seg_2 <- dna_seg(read.csv("~/master/rumen/dataflow/00-meta/4309680.csv")) 
df_seg_3 <- dna_seg(read.csv("~/master/rumen/dataflow/00-meta/CP022905.csv"))
df_seg_4 <- dna_seg(read.csv("~/master/rumen/dataflow/00-meta/KP339868.1_1.csv"))

dna_segs <- list(df_seg_1, df_seg_2, df_seg_3, df_seg_4)

names <- c("RUG782", "4309680", "CP022905", "KP339868.1_1")
names(dna_segs) <- names

comparison1 <- as.comparison(read.csv("~/master/rumen/dataflow/00-meta/RUG782v4309680.csv"))
comparison2 <- as.comparison(read.csv("~/master/rumen/dataflow/00-meta/4309680vCP022905.csv"))
comparison3 <- as.comparison(read.csv("~/master/rumen/dataflow/00-meta/CP022905vKP339868.1_1.csv"))

comparisons <- list(comparison1, comparison2, comparison3)

comparisons[[1]]$col <- apply_color_scheme(c(0.6, 0.4, 0.5, 0.6, 0.4, 0.5, 0.6, 0.4, 0.5, 0.6, 0.4, 0.5, 0.6), "grey")
comparisons[[2]]$col <- apply_color_scheme(c(0.6, 0.4, 0.5, 0.6, 0.4, 0.5, 0.6, 0.4, 0.5, 0.6), "grey")
comparisons[[3]]$col <- apply_color_scheme(c(0.6, 0.4, 0.5, 0.6, 0.4, 0.5, 0.6, 0.4, 0.5, 0.6), "grey")

#annot <- annotation(x1=c(130286, 130621, 139839), x2=c(130344, 131687, 139883), text=c("IR 1", "IR 2", "IR 3"), rot=c(0, 0, 0), col=c("black", "black", "black"))


plot_gene_map(dna_segs=dna_segs, comparisons=comparisons, xlims=xlims,  dna_seg_scale=TRUE, scale=FALSE)


#plot_gene_map(dna_segs=list(df_seg_1, df_seg_2), comparisons=list(comparison1), xlims=xlims, main="BH vs BQ, comparison of the first 50 kb", gene_type="side_blocks", dna_seg_scale=TRUE, scale=FALSE)
```

\newpage

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=3.5, fig.width=10}
library(genoPlotR)

xlims <- list(c(0, 45000), c(Inf, -Inf), c(Inf, -Inf),  c(Inf, -Inf))

df_seg_1 <- dna_seg(read.csv("~/master/rumen/dataflow/00-meta/RUG782.csv"))
df_seg_2 <- dna_seg(read.csv("~/master/rumen/dataflow/00-meta/4309680.csv")) 
df_seg_3 <- dna_seg(read.csv("~/master/rumen/dataflow/00-meta/CP022905.csv"))
df_seg_4 <- dna_seg(read.csv("~/master/rumen/dataflow/00-meta/KP339868.1_1.csv"))

dna_segs <- list(df_seg_4, df_seg_3, df_seg_2, df_seg_1)

names <- c("KP339868.1_1", "CP022905", "4309680", "RUG782")

names(dna_segs) <- names

comparison1 <- as.comparison(read.csv("~/master/rumen/dataflow/00-meta/KP339868.1_1vCP022905.csv"))
comparison2 <- as.comparison(read.csv("~/master/rumen/dataflow/00-meta/CP022905v4309680.csv"))
comparison3 <- as.comparison(read.csv("~/master/rumen/dataflow/00-meta/4309680vRUG782.csv"))

comparisons <- list(comparison1, comparison2, comparison3)

comparisons[[3]]$col <- apply_color_scheme(c(0.6, 0.4, 0.5, 0.6, 0.4, 0.5, 0.6, 0.4, 0.5, 0.6, 0.4, 0.5, 0.6), "grey")
comparisons[[2]]$col <- apply_color_scheme(c(0.6, 0.4, 0.5, 0.6, 0.4, 0.5, 0.6, 0.4, 0.5, 0.6), "grey")
comparisons[[1]]$col <- apply_color_scheme(c(0.6, 0.4, 0.5, 0.6, 0.4, 0.5, 0.6, 0.4, 0.5, 0.6), "grey")

#annot <- annotation(x1=c(130286, 130621, 139839), x2=c(130344, 131687, 139883), text=c("IR 1", "IR 2", "IR 3"), rot=c(0, 0, 0), col=c("black", "black", "black"))


plot_gene_map(dna_segs=dna_segs, comparisons=comparisons, xlims=xlims,  dna_seg_scale=TRUE, scale=FALSE)


#plot_gene_map(dna_segs=list(df_seg_1, df_seg_2), comparisons=list(comparison1), xlims=xlims, main="BH vs BQ, comparison of the first 50 kb", gene_type="side_blocks", dna_seg_scale=TRUE, scale=FALSE)
```


\newpage

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.height=7}
island_mapping_ava <- read.delim("~/master/rumen/dataflow/02-blast/resistance_island_mapping_allvall.txt", header=FALSE)
colnames(island_mapping_ava) <- c("qseqid", "sseqid", "pident", "sstart", "send", "qstart", "qend", "evalue", "bitscore", "score", "qlen", "length", "sseq")

island_mapping_ava$qseqid <- as.character(island_mapping_ava$qseqid)
island_mapping_ava$sseqid <- as.character(island_mapping_ava$sseqid)

island_mapping_ava <- island_mapping_ava %>%
  mutate(same = if_else(qseqid == sseqid, "yes", "no")) %>%
  filter(same != "yes") %>%
  filter(qseqid != "4309689-submission.assembly_79_resistance_genes") %>%
  filter(qseqid != "4309680-submission.assembly_52_resistance_genes") %>%
  filter(qseqid != "RUG782_1_resistance_genes") %>%
  mutate(same_seq = if_else(pident == 100 & qlen == length, "yes", "no")) %>%
  filter(same_seq == "yes") %>%
  group_by(qseqid) %>%
  mutate(group_num = length(unique(sseqid)) + 1) %>%
  ungroup() %>%
  select(qseqid, sseqid, group_num) %>%
  group_by(qseqid) %>%
  summarize(group_members = list(sort(c(sseqid, unique(qseqid))))) %>%
  ungroup() %>%
  select(-qseqid) %>%
  distinct()

list_groups <- unique(island_mapping_ava$group_members)
rep_seqs <- c()
num_in_group <- c()

for (i in 1:length(list_groups)){
  group_list <- list_groups[[i]]
  rnum <- sample(1:length(group_list), 1)
  rep_seqs <- c(rep_seqs, group_list[[rnum]])
  num_in_group <- c(num_in_group, length(group_list))
  
}

df_rep_seqs <- as.data.frame(cbind(rep_seqs, num_in_group))
colnames(df_rep_seqs)[1] <- "qseqid"

rem_seqs <- unlist(list_groups)[ !(unlist(list_groups) %in% df_rep_seqs$qseqid)]

island_mapping <- read.delim("~/master/rumen/dataflow/02-blast/resistance_island_mapping_allvall_V2.txt", header=FALSE)
colnames(island_mapping) <- c("qseqid", "sseqid", "pident", "sstart", "send", "qstart", "qend", "evalue", "bitscore", "score", "qlen", "length", "sseq")

df_agsinst_RUG782 <- island_mapping %>%
  select(qseqid, sseqid, pident, sstart, send, qstart, qend, bitscore, length) %>%
  filter(sseqid == "KP339868.1_1") %>%
  mutate(nstart = if_else(sstart < send, sstart, send)) %>%
  mutate(nsend = if_else(send > sstart, send, sstart)) %>%
  ungroup() %>%
  filter(qseqid != "KP339868.1_1") %>%
  group_by(qseqid) %>%
  mutate(bitscore_sum = sum(bitscore)) %>%
  ungroup() %>%
  filter(bitscore != 45953) %>%
  group_by(qseqid, sseqid) %>%
  mutate(length_sum = sum(length)) %>%
  ungroup() %>%
  filter(length_sum > 4000) %>%
  filter(pident > 80)

df_agsinst_RUG782_rumen <- df_agsinst_RUG782 %>%
  filter(qseqid == "4309689-submission.assembly_79_resistance_genes" | qseqid == "4309680-submission.assembly_52_resistance_genes" | qseqid == "RUG782_1_resistance_genes")
  

df_agsinst_RUG782 <- df_agsinst_RUG782 %>%
  full_join(df_rep_seqs) %>%
  bind_rows(df_agsinst_RUG782_rumen) %>%
  mutate(remove = qseqid %in% rem_seqs) %>%
  filter(remove != TRUE) 

df_agsinst_RUG782$newqseqid <- NA
df_agsinst_RUG782$num_in_group <- as.character(df_agsinst_RUG782$num_in_group)
df_agsinst_RUG782 <- as.data.frame(df_agsinst_RUG782)

for (j in 1:nrow(df_agsinst_RUG782)) {
  test <- df_agsinst_RUG782[j, "num_in_group"]
  if(is.na(test)) {
    df_agsinst_RUG782[j, "newqseqid"] <- df_agsinst_RUG782[j, "qseqid"]} else {
      df_agsinst_RUG782[j, "newqseqid"] <- paste(df_agsinst_RUG782[j, "qseqid"], " (", df_agsinst_RUG782[j, "num_in_group"], ")", sep = "")
    }
  }



df_agsinst_RUG782 <- df_agsinst_RUG782 %>%
  arrange(desc(bitscore_sum))
  

df_agsinst_RUG782$newqseqid <- gsub("-submission.assembly", "", df_agsinst_RUG782$newqseqid)
df_agsinst_RUG782$newqseqid <- gsub("_resistance_genes", "", df_agsinst_RUG782$newqseqid)

df_agsinst_RUG782$newqseqid <- factor(df_agsinst_RUG782$newqseqid, levels = unique(as.character(df_agsinst_RUG782$newqseqid)))

df_seq_species_type <- read.csv("~/master/rumen/dataflow/00-meta/resistance_island_seq_origins.csv")
df_seq_species_type$accession <- as.character(df_seq_species_type$accession)

df_agsinst_RUG782 <- df_agsinst_RUG782 %>%
  filter(qseqid != "AJ557257.1_1") %>%
  separate(qseqid, into = c("accession", "rm"), remove = FALSE, sep = "_") %>%
  inner_join(df_seq_species_type, by = c("accession")) %>%
  unite(newqseqid2, c("species", "type", "newqseqid"), sep = " | ", remove = FALSE)

df_agsinst_RUG782$newqseqid2 <- factor(df_agsinst_RUG782$newqseqid2, levels = unique(as.character(df_agsinst_RUG782$newqseqid2)))



library(ggplot2)
library(ggalt)
theme_set(theme_classic())

# gg <- ggplot(df_agsinst_RUG782, aes(x=nstart, xend=nsend, y=newqseqid, group=newqseqid)) + 
#         geom_dumbbell(size=2, aes(color = pident)) + 
#         #scale_x_continuous(label=percent) + 
#         labs(x=NULL, 
#              y=NULL, 
#              title="Dumbbell Chart") +
#         theme(plot.title = element_text(hjust=0.5, face="bold"),
#               plot.background=element_rect(fill="#f7f7f7"),
#               panel.background=element_rect(fill="#f7f7f7"),
#               panel.grid.minor=element_blank(),
#               panel.grid.major.y=element_blank(),
#               panel.grid.major.x=element_line(),
#               axis.ticks=element_blank(),
#               legend.position="top",
#               panel.border=element_blank()) 




gg <- ggplot(df_agsinst_RUG782, 
       aes(y = newqseqid2,
           x = nstart,
           xend = nsend)) +  
  geom_dumbbell(size = 1.5,
                size_x = 1.5, 
                size_xend = 1.5,
                colour_x = "red", 
                colour_xend = "red", aes(colour = pident)) +
  theme_minimal() + 
  labs(x = "Mapping to KP339868.1_1",
       y = "") +
    theme(text = element_text(size=8)) 
  

plot(gg)

df_unique_contigs <- df_agsinst_RUG782 %>%
  select(qseqid) %>%
  distinct()

#write.csv(df_unique_contigs, "~/master/rumen/dataflow/00-meta/resistance_blast_hit_cotigs_unique.csv")
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
df_ava <- read.delim("~/master/rumen/dataflow/02-blast/resistance_island_all_v_all_prot.txt", header=FALSE)
colnames(df_ava) <- c("qseqid", "sseqid", "pident", "sstart", "send", "qstart", "qend", "evalue", "bitscore", "score", "qlen", "length", "sseq")


df_header <- read.csv("~/master/rumen/dataflow/02-headers/resistance_island_blast_hits_concatenated_extractedCONTIGs_3rumen.csv")

df_header$file <- NULL

colnames(df_header) <- c("qseqid", "header")


df_header <- df_header %>%
  separate(header, into = c("rm", "start", "stop", "strand"), sep  = "#") %>%
  select(-rm)

df_header <- df_header[grepl("RUG782_1_resistance_genes", df_header$qseqid),]

df_header$start <- gsub(" ", "", df_header$start)
df_header$stop <- gsub(" ", "", df_header$stop)

df_header$start <- as.numeric(df_header$start)
df_header$stop <- as.numeric(df_header$stop)

df_ava <- df_ava[grepl("RUG782_1_resistance_genes", df_ava$qseqid),]
df_ava <- df_ava[!(grepl("RUG782_1_resistance_genes", df_ava$sseqid)),]



df_header$qseqid <- as.character(df_header$qseqid)
df_ava$qseqid <- as.character(df_ava$qseqid)

df_ava <- df_ava %>%
  filter(qlen == length) %>%
  inner_join(df_header) %>%
  filter(start > 110000) %>%
  filter(qseqid == "RUG782_1_resistance_genes_125" | qseqid == "RUG782_1_resistance_genes_126" | qseqid == "RUG782_1_resistance_genes_127" | qseqid == "RUG782_1_resistance_genes_128" | qseqid == "RUG782_1_resistance_genes_129" | qseqid == "RUG782_1_resistance_genes_130" | qseqid == "RUG782_1_resistance_genes_131" | qseqid == "RUG782_1_resistance_genes_132" | qseqid == "RUG782_1_resistance_genes_133" | qseqid == "RUG782_1_resistance_genes_134" | qseqid == "RUG782_1_resistance_genes_135" | qseqid == "RUG782_1_resistance_genes_136" | qseqid == "RUG782_1_resistance_genes_137" | qseqid == "RUG782_1_resistance_genes_138" | qseqid == "RUG782_1_resistance_genes_139") %>%
  group_by(qseqid) %>%
  mutate(mean_pident = mean(pident)) %>%
  mutate(coverage = length(unique(sseqid)))%>%
  ungroup() %>%
  select(qseqid, mean_pident, coverage) %>%
  distinct()

```



```{r, echo=FALSE, warning=FALSE, message=FALSE}
df_ribd <- read.delim("~/master/rumen/dataflow/02-blast/genomes_4_ribD.txt", header=FALSE)
colnames(df_ribd) <- c("qseqid", "sseqid", "pident", "sstart", "send", "qstart", "qend", "evalue", "bitscore", "score", "qlen", "length", "sseq")

df_ribd <- df_ribd  %>%
  filter(length > 180) %>%
  filter(bitscore > 100) %>%
  select(qseqid, sseqid) %>%
  distinct()

write.csv(df_ribd, "~/master/rumen/dataflow/00-meta/ribd_comparison.csv")

df_ribd <- read.delim("~/master/rumen/dataflow/02-blast/ref_seq_genomes_4_ribD.txt", header=FALSE)
colnames(df_ribd) <- c("qseqid", "sseqid", "pident", "sstart", "send", "qstart", "qend", "evalue", "bitscore", "score", "qlen", "length", "sseq")

df_ribd <- df_ribd  %>%
  filter(length > 180) %>%
  filter(bitscore > 100) %>%
  select(qseqid, sseqid) %>%
  distinct()

write.csv(df_ribd, "~/master/rumen/dataflow/00-meta/ribd_comparison_ref.csv")

df_ribd <- read.delim("~/master/rumen/dataflow/02-blast/rumen_genomes_4_ribD.txt", header=FALSE)
colnames(df_ribd) <- c("qseqid", "sseqid", "pident", "sstart", "send", "qstart", "qend", "evalue", "bitscore", "score", "qlen", "length", "sseq")

df_ribd <- df_ribd  %>%
  filter(length > 180) %>%
  filter(bitscore > 100) %>%
  select(qseqid, sseqid) %>%
  distinct()

write.csv(df_ribd, "~/master/rumen/dataflow/00-meta/ribd_comparison_rumen_genomes.csv")
```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
file <- "~/master/rumen/dataflow/01-prot/genomes_4_ribD_seqs.fasta"

df <- read.csv(paste(file, sep = ""), sep = ",", header = FALSE)

df$V1 <- as.character(df$V1)

headers <- df$V1[grepl(">", df$V1)]
seqs <- df$V1[!(grepl(">", df$V1))]

df <- as.data.frame(cbind(headers, seqs))

df$seqs <- gsub(" ", "", df$seqs)

df <- df %>%
  arrange(seqs) %>%
  rowwise() %>%
  mutate(len = nchar(seqs)) %>%
  distinct()

df <- as.data.frame(df)

df$headers <- gsub("_ribD", "", df$headers)

for (j in 1:nrow(df)){
  if (j <= nrow(df) - 1) {
  seq_down <- j + 1
  if (df[j,2] == df[seq_down,2]){
    df <- df[-j,]
  }
  }
}

df$headers <- as.character(df$headers)

ids <- unique(df$headers)

df_select_list <- list()
k <- 1

for (j in 1:nrow(df)){
      
      if (df[j, 3] > 300){
      df_select_list[[k]] <- paste(df[j, 1], "_LONG", sep ="")
      k <- k + 1 } else {
      df_select_list[[k]] <- paste(df[j, 1], "_SHORT", sep ="")
      k <- k + 1 
      }
      
      df_select_list[[k]] <- df[j, 2]
      k <- k + 1
}



dir_out <- "~/master/rumen/dataflow/01-prot/"

selected_long <- as.matrix(unlist(rbind(df_select_list)))
selected_long <- gsub("\"", "", selected_long)
write.table(selected_long, paste(dir_out, "genomes_4_ribD_seqs_unique.fasta"), row.names = FALSE, 
    col.names = FALSE, quote = FALSE)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
file <- "~/master/rumen/dataflow/01-prot/reference_genomes_ribD_seqs.fasta"

df <- read.csv(paste(file, sep = ""), sep = ",", header = FALSE)

df$V1 <- as.character(df$V1)

headers <- df$V1[grepl(">", df$V1)]
seqs <- df$V1[!(grepl(">", df$V1))]

df <- as.data.frame(cbind(headers, seqs))

df$seqs <- gsub(" ", "", df$seqs)

df <- df %>%
  arrange(seqs) %>%
  rowwise() %>%
  mutate(len = nchar(seqs)) %>%
  distinct()

df <- as.data.frame(df)

df$headers <- gsub("_ribD", "", df$headers)

for (j in 1:nrow(df)){
  if (j <= nrow(df) - 1) {
  seq_down <- j + 1
  if (df[j,2] == df[seq_down,2]){
    df <- df[-j,]
  }
  }
}

df$headers <- as.character(df$headers)


df_select_list <- list()
k <- 1

for (j in 1:nrow(df)){
      
      if (runif(1, 0, 1) < 0.5 ){
      
      if (df[j, 3] > 300){
      df_select_list[[k]] <- paste(">REF",j, "_LONG", sep ="")
      k <- k + 1 } else {
      df_select_list[[k]] <- paste(">REF",j, "_SHORT", sep ="")
      k <- k + 1 
      }
      
      df_select_list[[k]] <- df[j, 2]
      k <- k + 1

      } else {}
  }



dir_out <- "~/master/rumen/dataflow/01-prot/"

selected_long <- as.matrix(unlist(rbind(df_select_list)))
selected_long <- gsub("\"", "", selected_long)
write.table(selected_long, paste(dir_out, "reference_genomes_ribD_seqs_unique.fasta"), row.names = FALSE, 
    col.names = FALSE, quote = FALSE)



```

```{r, echo=FALSE, warning=FALSE, message=FALSE}
file <- "~/master/rumen/dataflow/01-prot/rumen_genomes_ribD_seqs.fasta"

df <- read.csv(paste(file, sep = ""), sep = ",", header = FALSE)

df$V1 <- as.character(df$V1)

headers <- df$V1[grepl(">", df$V1)]
seqs <- df$V1[!(grepl(">", df$V1))]

df <- as.data.frame(cbind(headers, seqs))

df$seqs <- gsub(" ", "", df$seqs)

df <- df %>%
  arrange(seqs) %>%
  rowwise() %>%
  mutate(len = nchar(seqs)) %>%
  distinct() 

df <- as.data.frame(df)

df$headers <- gsub("_ribD", "", df$headers)

for (j in 1:nrow(df)){
  if (j <= nrow(df) - 1) {
  seq_down <- j + 1
  if (df[j,2] == df[seq_down,2]){
    df <- df[-j,]
  }
  }
}

df$headers <- as.character(df$headers)

df <- df[!(df$headers %in% ids),]

df_select_list <- list()
k <- 1

for (j in 1:nrow(df)){
  
      
  if (runif(1, 0, 1) < 0.1 ){    
  
      if (df[j, 3] > 300){
      df_select_list[[k]] <- paste(">RUMEN", j, "_LONG", sep ="")
      k <- k + 1 } else {
      df_select_list[[k]] <- paste(">RUMEN", j, "_SHORT", sep ="")
      k <- k + 1 
      }
      
      df_select_list[[k]] <- df[j, 2]
 
      k <- k + 1
      
      } else {}
}



dir_out <- "~/master/rumen/dataflow/01-prot/"

selected_long <- as.matrix(unlist(rbind(df_select_list)))
selected_long <- gsub("\"", "", selected_long)
write.table(selected_long, paste(dir_out, "rumen_genomes_ribD_seqs_unique.fasta"), row.names = FALSE, 
    col.names = FALSE, quote = FALSE)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
file <- "~/master/rumen/dataflow/01-prot/rumen_genomes_ribD_seqs.fasta"

df <- read.csv(paste(file, sep = ""), sep = ",", header = FALSE)

df$V1 <- as.character(df$V1)

headers <- df$V1[grepl(">", df$V1)]
seqs <- df$V1[!(grepl(">", df$V1))]

df <- as.data.frame(cbind(headers, seqs))

df$seqs <- gsub(" ", "", df$seqs)

df <- df %>%
  arrange(seqs) %>%
  rowwise() %>%
  mutate(len = nchar(seqs)) %>%
  distinct() %>%
  filter(len < 250)

df <- as.data.frame(df)

df$headers <- gsub("_ribD", "", df$headers)

for (j in 1:nrow(df)){
  if (j <= nrow(df) - 1) {
  seq_down <- j + 1
  if (df[j,2] == df[seq_down,2]){
    df <- df[-j,]
  }
  }
}

df$headers <- as.character(df$headers)

df <- df[!(df$headers %in% ids),]

df_select_list <- list()
k <- 1

for (j in 1:nrow(df)){
  
      
  if (runif(1, 0, 1) < 0.1 ){    
  
      if (df[j, 3] > 300){
      df_select_list[[k]] <- paste(">RUMEN", j, "_LONG", sep ="")
      k <- k + 1 } else {
      df_select_list[[k]] <- paste(">RUMEN", j, "_SHORT", sep ="")
      k <- k + 1 
      }
      
      df_select_list[[k]] <- df[j, 2]
 
      k <- k + 1
      
      } else {}
}



dir_out <- "~/master/rumen/dataflow/01-prot/"

selected_long <- as.matrix(unlist(rbind(df_select_list)))
selected_long <- gsub("\"", "", selected_long)
write.table(selected_long, paste(dir_out, "rumen_genomes_ribD_seqs_unique_short.fasta"), row.names = FALSE, 
    col.names = FALSE, quote = FALSE)
```


```{r, echo=FALSE, warning=FALSE, message=FALSE}
df_path <- read.csv("~/master/rumen/dataflow/02-headers/genomes_4_ribD.csv")
df_path$file <- NULL

colnames(df_path) <- c("gene", "header")

df_path <- df_path %>%
  separate(header, into = c("rm", "start", "end", "strand"), sep = "#") %>%
  select(-rm)

df_path$gene <- as.character(df_path$gene)

rumen1 <- c("4309680-submission.assembly_35_14",
"4309680-submission.assembly_35_15",
"4309680-submission.assembly_35_16",
"4309680-submission.assembly_35_17",
"4309680-submission.assembly_35_18",
"4309680-submission.assembly_35_19",
"4309680-submission.assembly_35_20",
"4309680-submission.assembly_35_21",
"4309680-submission.assembly_35_22",
"4309680-submission.assembly_35_23",
"4309680-submission.assembly_35_24",
"4309680-submission.assembly_35_25")

df_4309680 <- df_path[df_path$gene %in% rumen1, ]

#write.csv(df_4309680 , "~/master/rumen/dataflow/00-meta/4309680_RibD.csv")

pathogen1 <- c("CP022905.1_1129",
"CP022905.1_1130",
"CP022905.1_1131",
"CP022905.1_1132",
"CP022905.1_1133",
"CP022905.1_1134",
"CP022905.1_1135",
"CP022905.1_1136",
"CP022905.1_1137",
"CP022905.1_1138",
"CP022905.1_1139",
"CP022905.1_1140")

df_CP022905 <- df_path[df_path$gene %in% pathogen1, ]

#write.csv(df_CP022905 , "~/master/rumen/dataflow/00-meta/CP022905_RibD.csv")

```

trimming and then remove seqs that don't fall in a clear cluster

```{r, echo=FALSE, warning=FALSE, message=FALSE}
df_ant6 <- read.delim("~/master/rumen/dataflow/02-blast/3_resistance_genes.txt", header=FALSE)
colnames(df_ant6) <- c("qseqid", "sseqid", "pident", "sstart", "send", "qstart", "qend", "evalue", "bitscore", "score", "qlen", "length", "sseq")


df_ant6 <- df_ant6 %>%
  filter(qseqid == "4309680-submission.assembly_52_ANT6")

write.csv(df_ant6, "~/master/rumen/dataflow/00-meta/ANT6_rumen.csv")  

#

df_aph3 <- read.delim("~/master/rumen/dataflow/02-blast/3_resistance_genes.txt", header=FALSE)
colnames(df_aph3) <- c("qseqid", "sseqid", "pident", "sstart", "send", "qstart", "qend", "evalue", "bitscore", "score", "qlen", "length", "sseq")


df_aph3 <- df_aph3 %>%
  filter(qseqid == "4309680-submission.assembly_52_APH3")

write.csv(df_aph3, "~/master/rumen/dataflow/00-meta/APH3_rumen.csv")  

#

df_sat4 <- read.delim("~/master/rumen/dataflow/02-blast/3_resistance_genes.txt", header=FALSE)
colnames(df_sat4) <- c("qseqid", "sseqid", "pident", "sstart", "send", "qstart", "qend", "evalue", "bitscore", "score", "qlen", "length", "sseq")


df_sat4 <- df_sat4 %>%
  filter(qseqid == "4309680-submission.assembly_52_sat4")  

write.csv(df_sat4, "~/master/rumen/dataflow/00-meta/SAT4_rumen.csv")  
```
