---
output:
  pdf_document: default
  html_document: default
---

```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(ggthemes)
library(printr)
library(ggplot2)
library(cowplot)
library(reshape2)
library(knitr)
library(gmodels)
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
### METADATA
### ONLY LOOK AT BACTERIA PRIMER

df_meta <- read.csv("~/master/rumen/dataflow/00-meta/henderson2015_fastq.csv")

df_meta <- df_meta %>%
  select(Experiment, LibraryName, GRCid, Simple.classification..animal., Age.classification..animal., Gender..animal., Forage.proportion..diet., Concentrate.proportion..diet., Forage.concentrate.category..diet., Starch.rich..diet., Pectin.rich...diet., Weeks.on.diet..diet.)

colnames(df_meta) <- c("sra_accession", "lib_name", "GRCid", "animal_class", "animal_age", "gender", "forage_proportion", "concentrate_proportion", "diet_category", "starch_rich", "pectin_rich", "week_on_diet")

df_meta <- df_meta %>%
  separate(lib_name, into = c("rm", "primer"), sep = "\\.") %>%
  rowwise() %>%
  mutate(bacteria_primer = substr(primer, 1, 1) == "B") %>%
  filter(bacteria_primer == TRUE) %>%
  select(-rm)
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
### CLASSIFICATION
### ONLY SEQUENCES CLASSIFIED AT PHYLUM LEVEL

df_classification  <- read.csv("~/master/rumen/dataflow/03-asv-taxonomy/henderson2015-20_320-99-rdp.csv")
df_classification[,1] <- NULL

df_classification <- df_classification %>%
  select(asv_id, phylum, family, genus) %>%
  filter(phylum != "NA")
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
### OTU TABLE

df <- read.delim("~/master/rumen/dataflow/03-asv-table/henderson2015-20_320-99.txt", skip = 1, header = TRUE)
df$clustering <- 99
colnames(df)[1] <- 'asv_id'

col_to_gather <- names(df)[!(startsWith(names(df), "SRX"))]
df <- melt(df, id = col_to_gather)
colnames(df)[3:4] <- c('sra_accession', 'count')

df$asv_id <- as.character(df$asv_id)
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
### SAMPLES ONLY WITH MORE THAN 1000 READS

df_normalization <- df %>%
  filter(clustering == 99) %>%
  group_by(sra_accession) %>%
  mutate(total_reads = sum(count)) %>%
  ungroup() %>%
  select(sra_accession, total_reads) %>%
  ungroup() %>%
  distinct() %>%
  filter(total_reads > 500)
  

df_complete <- inner_join(df, df_meta) %>%
  inner_join(df_normalization) %>%
  inner_join(df_classification)

rm(list=setdiff(ls(), "df_complete"))
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_lacto_positive <- df_complete %>%
  filter(genus == "Lactobacillus") %>%
  filter(count != 0)


lacto_positive_samples <- unique(df_lacto_positive$GRCid)

df_complete_starch <- df_complete %>%
  filter(starch_rich == "y") 


df_complete_starch <- df_complete_starch%>%
  rowwise() %>%
  mutate(lacto_signal = ifelse(GRCid %in% lacto_positive_samples, "pos", "neg")) %>%
  ungroup() 
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_metrics <- df_complete_starch %>% 

  mutate(count_norm = (count / total_reads)*100) %>%
  
  unite(classification, c("phylum", "family", "genus"), sep = ";", remove = FALSE) %>%
  
  distinct() 
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_phlycounts_counts <- df_metrics %>%
  select(asv_id, GRCid, count) %>%
  spread(GRCid, count)

df_phlycounts_counts[is.na(df_phlycounts_counts)] <- 0

df_phlycounts_counts <- df_phlycounts_counts[order(df_phlycounts_counts$asv_id),] 

numsamples <- length(df_phlycounts_counts)

otumat <- as.matrix(df_phlycounts_counts[,2:numsamples])
rownames(otumat) <- as.data.frame(df_phlycounts_counts)[,1]

df_phlycounts_tax <- df_metrics %>%
  select(asv_id, phylum, family, genus) %>%
  unique()

df_phlycounts_tax$phylum <- as.character(df_phlycounts_tax$phylum)
df_phlycounts_tax$family <- as.character(df_phlycounts_tax$family)
df_phlycounts_tax$genus <- as.character(df_phlycounts_tax$genus)

df_phlycounts_tax[is.na(df_phlycounts_tax)] <- "NotAssigned"

df_phlycounts_tax <- df_phlycounts_tax[order(df_phlycounts_tax$asv_id),] 

taxmat <- as.matrix(df_phlycounts_tax[,2:4])
rownames(taxmat) <- as.data.frame(df_phlycounts_tax)[,1]

df_phylocounts_meta <- df_metrics %>% 
  select(GRCid, lacto_signal) %>%
  unique() 

df_phylocounts_meta$lacto_signal <- as.factor(df_phylocounts_meta$lacto_signal)

df_phylocounts_meta <- as.data.frame(df_phylocounts_meta)

row.names(df_phylocounts_meta) <- df_phylocounts_meta[,1]
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
library("phyloseq")
library("ape")
library("DESeq2")

OTU = otu_table(otumat, taxa_are_rows = TRUE)
TAX = tax_table(taxmat)
physeq = phyloseq(OTU, TAX)

samplesdata <- sample_data(df_phylocounts_meta)

physeq = phyloseq(OTU, TAX, samplesdata)

diagdds = phyloseq_to_deseq2(physeq, ~ lacto_signal)

#calculate geometric means prior to estimate size factors
gm_mean = function(x, na.rm=TRUE){
  exp(sum(log(x[x > 0]), na.rm=na.rm) / length(x))
}

geoMeans = apply(counts(diagdds), 1, gm_mean)
diagdds = estimateSizeFactors(diagdds, geoMeans = geoMeans)

rm(list=setdiff(ls(), c("diagdds", "physeq", "df_metrics")))

diagdds = DESeq(diagdds, fitType = "parametric", test = "Wald", parallel=TRUE)

res = results(diagdds, cooksCutoff = FALSE)
alpha = 0.05
#sigtab = res[which(res$padj < alpha), ]
sigtab = res[which(res$pvalue < alpha), ]

sigtab = cbind(as(sigtab, "data.frame"), as(tax_table(physeq)[rownames(sigtab), ], "matrix")) 

selected_seq_ids <- row.names(sigtab)

df_plot <- df_metrics[df_metrics$asv_id %in% selected_seq_ids,]

df_direction <- as.data.frame(sigtab[,2]) 
colnames(df_direction) <- "log_fold_change"
df_direction$asv_id <- selected_seq_ids

df_direction <- df_direction %>%
  rowwise() %>%
  mutate(direction = ifelse(log_fold_change > 0, "increase", "decrease"))

df_plot <- df_plot %>% 
  inner_join(df_direction) %>% 
  mutate(fold_change = 2 * log_fold_change) %>%
  rowwise() %>%
  mutate(above1 = ifelse(count_norm > 2, 1, 0)) %>%
  group_by(asv_id) %>%
  mutate(sum_above1 = sum(above1)) %>%
  ungroup() %>%
  filter(sum_above1 > 5) %>%
  filter(family == "Prevotellaceae") %>%
  filter(fold_change > 5 | fold_change < -5)
```

\newpage

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=20, fig.height=10}
ggplot(df_plot, aes(x=lacto_signal, y=count_norm)) + 
            geom_point(aes(colour = lacto_signal), size = 5) + 
  facet_wrap( ~ asv_id, scales="free") +
  theme() +
          theme(strip.text = element_text(size = 40),
          plot.title = element_text(size = 30),
          axis.text.x = element_text(angle = 90, hjust = 1, size = 20),
          axis.text.y = element_text(size = 20),
          axis.title.x = element_text(size = 20),
          axis.title.y = element_text(size = 20))
```



```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_save <- df_plot %>%
  select(asv_id, direction, fold_change) %>%
  distinct() 

write.csv(df_save, "~/master/rumen/dataflow/00-meta/lacto_signal_differential.csv")
```