---
output:
  pdf_document: default
  html_document: default
---

```{r, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(ggthemes)
library(printr)
library(ggplot2)
library(cowplot)
library(reshape2)
library(knitr)
library(pwr)
library("ggpubr")
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
two_way_anova_extract <- function(df, value = 1, y = "NA", factor1 = "NA", factor2 = "NA"){
  ### Function that carries out a two-way anova and gives back the first, second or third (interaction) p-value based on hat is specified. Dataframe must have each factor be a column and the dependant variable be the third column.
  
  df <- as.data.frame(df)
  # rename the columns to generalize the method
  colnames(df)[which(names(df) == y)] <- "dependant"
  colnames(df)[which(names(df) == factor1)] <- "category1"
  colnames(df)[which(names(df) == factor2)] <- "category2"
  
  res.aov2 <- aov(dependant ~ category1 * category2, data = df)
  sum <- summary(res.aov2)
  
  # this would get the name from the summary object
  #test_name <- gsub(" ", "", rownames(sum[[1]])[value])
  
  # assign name to pvalue
  if(value == 1){
    test_name <- factor1
  } else if (value == 2){
    test_name <- factor2
  } else if (value == 3) {
    test_name <- paste(factor1, factor2, sep = ":")
  }
  
  # extract p value
  pvalue <- sum[[1]][["Pr(>F)"]][value]
  
  # only return p value for now 
  return(pvalue)
}

```


## plot of sequence quality

- Plotting the median phred quality score from the sequencing. I typically find that you don't want to go below a quality score of 35, which presents somewhere between 99.9 and 99.99% accuracy.  
- It looks like these were reads that were already collapsed from the paried-end sequences.  
- The is a massive drop in quality in the middle of the sequence, so I will trim from 0 to 200 basepairs, as this is probably the only safe region to work with.  

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=20, fig.height=5}
quality_qiime <- read.csv("~/master/rumen/dataflow/02-qiime/wetzels2018-seq-quality-summary.csv", header=FALSE)
quality_qiime <- as.data.frame(t(quality_qiime))

quality_qiime[, ] <- lapply(quality_qiime[, ], as.character)


quality_qiime[1,] <- gsub("\\%", "", quality_qiime[1,])

colnames(quality_qiime) <- quality_qiime[1,]

quality_qiime <- quality_qiime[-1,]

colnames(quality_qiime)[1] <- "position"

quality_qiime$'50' <- as.numeric(quality_qiime$'50')

xbreaks <- quality_qiime$position[seq(1, length(quality_qiime$position), 10)]

qplot(quality_qiime$position, quality_qiime$'50') + 
  theme(axis.text.x = element_text(angle = 90, hjust = 1, size = 18),
        axis.text.y = element_text(size = 18), 
        axis.title.x = element_text(size = 25), 
        axis.title.y = element_text(size = 25)) +
  scale_x_discrete("Position",  breaks = xbreaks) +
  ylab("Median Quality Score")
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_meta <- read.csv("~/master/rumen/dataflow/00-meta/wetzels2018_fastq.csv")
df_classification <- read.csv("~/master/rumen/dataflow/03-asv-taxonomy/wetzels2017-1_195-100-rdp.csv")
df_classification[,1] <- NULL

df_100 <- read.delim("~/master/rumen/dataflow/03-asv-table/wetzels2017-1_195-100.txt", skip = 1, header = TRUE)
df_100$clustering <- 100

df_99 <- read.delim("~/master/rumen/dataflow/03-asv-table/wetzels2017-1_195-99.txt", skip = 1, header = TRUE)
df_99$clustering <- 99

df_97 <- read.delim("~/master/rumen/dataflow/03-asv-table/wetzels2017-1_195-97.txt", skip = 1, header = TRUE)
df_97$clustering <- 97

df <- bind_rows(df_100, df_99, df_97)

colnames(df)[1] <- 'asv_id'

col_to_gather <- names(df)[!startsWith(names(df), "ERX")]
df <- melt(df, id = col_to_gather)
colnames(df)[3:4] <- c('sra_accession', 'count')

df$asv_id <- as.character(df$asv_id)
df_meta$day <- as.character(df_meta$day)
df_meta$treatment <- as.character(df_meta$treatment)

df_normalization <- df %>%
  filter(clustering == 100) %>%
  group_by(sra_accession) %>%
  mutate(total_16s_reads = sum(count)) %>%
  ungroup() %>%
  select(sra_accession, total_16s_reads) %>%
  ungroup() %>%
  distinct() 

df_classification <- df_classification %>%
  select(asv_id, phylum, family, genus)


df_complete <- inner_join(df, df_meta) %>%
  inner_join(df_normalization) %>%
  inner_join(df_classification)
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_metrics <- df_complete %>% 
  
  filter(count > 0) %>%
  
  mutate(count_norm = (count / total_16s_reads)*100) %>%
  
  group_by(asv_id, clustering, day, treatment) %>%
  mutate(med_count_norm = median(count_norm)) %>%
  mutate(sd_count_norm = sd(count_norm)) %>%
  mutate(cv_count_norm =  (sd(count_norm) / mean(count_norm))*100) %>%
  mutate(n_count = length(count_norm)) %>%
  ungroup() %>%
  
  filter(n_count == 3) %>%
  
  group_by(asv_id, clustering) %>%
  mutate(ndays = length(unique(day))) %>%
  mutate(ntreatment = length(unique(treatment))) %>%
  ungroup() %>%
  
  filter(ndays > 2) %>%
  
  filter(ntreatment == 2) %>%
  
  unite(classification, c("phylum", "family", "genus"), sep = ";")

df_metrics$classification <- gsub(";NA", "", df_metrics$classification)
```

## data clean up

- Data with no counts for a sequence were removed.  
- Data without triplicates remaining were also removed.  
- Data without more than 2 days and both treatments were removed.  

## plot of approximate abundances

- Percent of the sample reaches up to 12%. 
- A good percentage of the data lies between 0.5 and 3%. 

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=7.5, fig.height=3}
df_abundance <- df_metrics %>%
  select(asv_id, clustering, med_count_norm) %>%
  unique() %>%
  arrange(med_count_norm)

df_abundance$clustering <- as.factor(df_abundance$clustering)

plot <- ggplot(df_abundance, aes(x=reorder(asv_id, med_count_norm))) +
    theme_few() +
    geom_point(aes(y = med_count_norm, colour = clustering),
      stat = "identity", fill = "lightgrey", size = 0.75) +
    theme(strip.text = element_text(size = 20),
          axis.text.y = element_text(size = 20), 
          axis.title.y = element_text(size = 20)) +
    theme(axis.title.x=element_blank(),
        axis.text.x=element_blank(),
        axis.ticks.x=element_blank()) +
  facet_grid(. ~ clustering) +
  theme(legend.position="none") +
  ylab("Percent of Sample")
plot
```

## clostridium box plot for infected sample 

- The pathogen seemed to be added after day 7 and was perhaps quite stable.  

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE, fig.width=20, fig.height=6}
df_clostridium <- df_complete %>%
  filter(asv_id == "4b16c075d92bebc14342feb2d42b233a") %>%
  filter(treatment == "non-infected") %>%
  
  mutate(count_norm = (count / total_16s_reads)*100)


ggboxplot(df_clostridium, x = "day", y = "count_norm", color = "treatment",
          palette = c("#00AFBB", "#E7B800")) +
          ylab("Percent of Sample") + 
          xlab("Day") +
          theme(strip.text = element_text(size = 20),
          plot.title = element_text(size = 20), 
          axis.text.x = element_text(angle = 90, hjust = 1, size = 20),  
          axis.text.y = element_text(size = 20), 
          axis.title.x = element_text(size = 20), 
          axis.title.y = element_text(size = 20)) + 
  
          facet_grid(. ~ clustering)
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
# blogged followed to carry out the ANOVA and Pairwise test
# http://www.sthda.com/english/wiki/two-way-anova-test-in-r
```

## 2 way anova and box plots

- Carried out a two way ANOVA. 
- Selected data with a p-value lower than 0.05 for the treatment and treatment day interaction. Also ensured that the resulted significant sequences had a median percent of over 0.5 for the experiment. These plots are to show any differences between control and treatment.
- Plots sorted by overall aundance. 

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_metrics$day <- factor(df_metrics$day, levels = c("5", "6", 
  "7", "10", "12", "15"))
df_metrics$treatment <- as.factor(df_metrics$treatment)
df_anova_list <- list()
i <- 1

for (seq in unique(df_metrics$asv_id)) {
  
  df_anova_select1 <- df_metrics %>% 
    filter(day == "10" | day == "12" | day == "15") %>%
    select(asv_id, clustering, day, treatment, count_norm) %>% 
    filter(asv_id == seq)
  
  for (cluster in unique(df_anova_select1$clustering)) {
    
    
    df_anova_select2 <- df_anova_select1 %>% 
      filter(clustering == cluster)
      
  if (length(unique(df_anova_select2$day)) > 1 & length(unique(df_anova_select2$treatment)) > 1){
    
      df_anova_select2$treatment_anova <- two_way_anova_extract(df_anova_select2, 
        value = 1, y = "count_norm", factor1 = "treatment", 
        factor2 = "day")
      df_anova_select2$day_anova <- two_way_anova_extract(df_anova_select2, 
        value = 2, y = "count_norm", factor1 = "treatment", 
        factor2 = "day")
      df_anova_select2$treatment_day_anova <- two_way_anova_extract(df_anova_select2, 
        value = 3, y = "count_norm", factor1 = "treatment", 
        factor2 = "day")
      
      df_anova_list[[i]] <- df_anova_select2
      i <- i + 1
}
  }
}

df_anova <- bind_rows(df_anova_list) %>%
  select(-count_norm) %>%
  distinct()
```


```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_treatment_plot <- inner_join(df_metrics, df_anova) %>%
  
  filter(treatment_anova < 0.05) %>%
  filter(treatment_day_anova < 0.05) %>%

  group_by(asv_id, clustering) %>%
  mutate(exp_med = median(count_norm)) %>%
  ungroup() %>%
  
  arrange(desc(count_norm)) %>%
  
  filter(classification != "NA") %>%
  
  filter(exp_med > 0.5)

# dummy data frames to fix facet in plotting. 
df_dummy <- df_treatment_plot %>%
  select(clustering, sample_id, day) %>%
  unique()

df_dummy_100 <- df_dummy  %>%
  filter(clustering == 100)

df_dummy_99 <- df_dummy  %>%
  filter(clustering == 99)

df_dummy_97 <- df_dummy  %>%
  filter(clustering == 97)

seqs <- unique(df_treatment_plot$asv_id)
plots <- list()
i <- 1

for (seq in seqs){
  
    df_p <- df_treatment_plot %>%
      filter(asv_id == seq)
  
    if( !(100 %in% unique(df_p$clustering)) ){
      df_p <- bind_rows(df_p, df_dummy_100 )
    }
    
    if( !(99 %in% unique(df_p$clustering)) ){
      df_p <- bind_rows(df_p, df_dummy_99 )
    }
    
    if( !(97 %in% unique(df_p$clustering)) ){
      df_p <- bind_rows(df_p, df_dummy_97 )
    }
      
    taxa <- unique(df_p$classification)
    
plot <- ggboxplot(df_p, x = "day", y = "count_norm", color = "treatment",
          palette = c("#00AFBB", "#E7B800")) +
          ylab("Percent of Sample") + 
          xlab("Day") +
          theme(strip.text = element_text(size = 20),
          plot.title = element_text(size = 20), 
          axis.text.x = element_text(angle = 90, hjust = 1, size = 20),  
          axis.text.y = element_text(size = 20), 
          axis.title.x = element_text(size = 20), 
          axis.title.y = element_text(size = 20)) + 
  
          facet_grid(. ~ clustering) + 
          ggtitle(taxa)
    

plots[[i]] <- plot
    
    i <- i + 1
}
```

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=20, fig.height=6}
for (i in 1:length(plots)) {
  plot(plots[[i]])
  cat("\r\n\r\n")
}
```

## carried out a pairwise t-test with correction for control data over time

- Data selected where there were more than 2 significant tests, using a p value of 0.005, over the time course.
- Plots sorted by overall aundance. 

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_pairpval_list <- list()
i <- 1

for (seq in unique(df_metrics$asv_id)) {
  
  df_pairpval_select1 <- df_metrics %>% 
    filter(treatment == "non-infected") %>%
    select(asv_id, clustering, day, count_norm) %>% 
    filter(asv_id == seq)
  
  for (cluster in unique(df_pairpval_select1$clustering)) {
    
    df_pairpval_select2 <- df_pairpval_select1 %>% 
      filter(clustering == cluster)
      
    pairwise_test <- pairwise.t.test(df_pairpval_select2$count_norm, df_pairpval_select2$day, p.adjust.method = "BH")
    
    pairwise_values <- c(pairwise_test$p.value)
    
    pairwise_values  <- pairwise_values[!is.na(pairwise_values)]
    
    df_pairpval_select2$count_sig_ttests <- sum(pairwise_values < 0.01)
    
    df_pairpval_list[[i]] <- df_pairpval_select2 
    i <- i + 1

  }
}

df_pairpval <- bind_rows(df_pairpval_list) %>%
  select(-count_norm) %>%
  distinct()
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_days_plot <- inner_join(df_metrics, df_pairpval) %>%
  filter(count_sig_ttests > 2) %>%
  
  group_by(asv_id, clustering) %>%
  mutate(exp_med = median(count_norm)) %>%
  ungroup() %>%
  
  arrange(desc(count_norm))

seqs_changes_overtime <- unique(df_days_plot$asv_id)

seqs <- unique(df_days_plot$asv_id)
plots <- list()
i <- 1

for (seq in seqs){
  
    df_p <- df_days_plot %>%
      filter(asv_id == seq)
  
    if( !(100 %in% unique(df_p$clustering)) ){
      df_p <- bind_rows(df_p, df_dummy_100 )
    }
    
    if( !(99 %in% unique(df_p$clustering)) ){
      df_p <- bind_rows(df_p, df_dummy_99 )
    }
    
    if( !(97 %in% unique(df_p$clustering)) ){
      df_p <- bind_rows(df_p, df_dummy_97 )
    }

    taxa <- unique(df_p$classification)
    
plot <- ggline(df_p, x = "day", y = "count_norm", color = "treatment",
          add = c("mean_se", "dotplot"),
          palette = c("#00AFBB", "#E7B800"), 
          facet.by = "clustering") +
          ylab("Percent of Sample") + 
          xlab("Day") +
          theme(strip.text = element_text(size = 20),
          plot.title = element_text(size = 20), 
          axis.text.x = element_text(angle = 90, hjust = 1, size = 20),  
          axis.text.y = element_text(size = 20), 
          axis.title.x = element_text(size = 20), 
          axis.title.y = element_text(size = 20)) + 
  
          #facet_grid(. ~ clustering) + 
          ggtitle(taxa)
    

plots[[i]] <- plot
    
    i <- i + 1
}
```

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=20, fig.height=6}
for (i in 1:length(plots)) {
  plot(plots[[i]])
  cat("\r\n\r\n")
}
```

\newpage

## change in abundant OTOs over time 

- Count of unique OTUs that are more than 1% of their respective dataset

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_diversity <- df_metrics %>%
  filter(day != 10) %>%
  filter(count_norm > 1) %>%
  filter(treatment == "non-infected") %>%
  select(asv_id, clustering, day, count_norm) %>%
  group_by(clustering, day) %>%
  mutate(n_otus = length(unique(asv_id))) %>%
  ungroup() %>%
  
  select(-asv_id, -count_norm) %>%
  distinct()

df_diversity$day <- as.numeric(as.character(df_diversity$day))

plot <- ggplot(df_diversity, aes(x=day, y=n_otus)) +
    theme_few() +
    geom_point(aes(y = n_otus),
      stat = "identity", fill = "lightgrey", size = 3) +
    theme(strip.text = element_text(size = 15),
          axis.text.y = element_text(size = 15), 
          axis.title.y = element_text(size = 15)) +
  facet_grid(. ~ clustering) +
  geom_smooth(method='lm') +
  ylab("Number of OTUs > 1%")

plot


```

\newpage

## change in low abundant OTOs over time 

- Count of unique OTUs that are less than 1% of their respective dataset

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_diversity <- df_metrics %>%
  filter(day != 10) %>%
  filter(count_norm < 1) %>%
  filter(treatment == "non-infected") %>%
  select(asv_id, clustering, day, count_norm) %>%
  group_by(clustering, day) %>%
  mutate(n_otus = length(unique(asv_id))) %>%
  ungroup() %>%
  
  select(-asv_id, -count_norm) %>%
  distinct()

df_diversity$day <- as.numeric(as.character(df_diversity$day))

plot <- ggplot(df_diversity, aes(x=day, y=n_otus)) +
    theme_few() +
    geom_point(aes(y = n_otus),
      stat = "identity", fill = "lightgrey", size = 3) +
    theme(strip.text = element_text(size = 15),
          axis.text.y = element_text(size = 15), 
          axis.title.y = element_text(size = 15)) +
  facet_grid(. ~ clustering) +
  geom_smooth(method='lm') +
  ylab("Number of OTUs < 1%")

plot

```

\newpage

## plotting of abundant OTUs over time

- Plotting all OTUs overtime with a median abundance of over 1% across the whole experiment

\vspace{12pt}

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_abundant <- df_metrics %>% 
    filter(treatment == "non-infected") %>%
    select(asv_id, clustering, day, count_norm) %>% 
    group_by(asv_id, clustering) %>%
    mutate(experiment_med = median(count_norm)) %>%
    ungroup() %>%
    filter(experiment_med > 1)
```

```{r, echo=FALSE, message=FALSE, warning=FALSE}
df_days_plot <- inner_join(df_metrics, df_abundant) %>%
  arrange(desc(experiment_med))

seqs <- unique(df_days_plot$asv_id)
plots <- list()
i <- 1

for (seq in seqs) {
  
    df_p <- df_days_plot %>%
      filter(asv_id == seq)
  
    if( !(100 %in% unique(df_p$clustering)) ){
      df_p <- bind_rows(df_p, df_dummy_100 )
    }
    
    if( !(99 %in% unique(df_p$clustering)) ){
      df_p <- bind_rows(df_p, df_dummy_99 )
    }
    
    if( !(97 %in% unique(df_p$clustering)) ){
      df_p <- bind_rows(df_p, df_dummy_97 )
    }
  
    taxa <- unique(df_p$classification)
    
if(seq %in% seqs_changes_overtime){
    
plot <- ggline(df_p, x = "day", y = "count_norm", color = "treatment",
          add = c("mean_se", "dotplot"),
          palette = c("#00AFBB", "#E7B800"), 
          facet.by = "clustering") +
          ylab("Percent of Sample") + 
          xlab("Day") +
          theme(strip.text = element_text(size = 20),
          plot.title = element_text(size = 20), 
          axis.text.x = element_text(angle = 90, hjust = 1, size = 20),  
          axis.text.y = element_text(size = 20), 
          axis.title.x = element_text(size = 20), 
          axis.title.y = element_text(size = 20)) + 
  
          #facet_grid(. ~ clustering) + 
          ggtitle(paste(taxa, " ***"    ))
    
} else {
  
  
plot <- ggline(df_p, x = "day", y = "count_norm", color = "treatment",
          add = c("mean_se", "dotplot"),
          palette = c("#00AFBB", "#E7B800"), 
          facet.by = "clustering") +
          ylab("Percent of Sample") + 
          xlab("Day") +
          theme(strip.text = element_text(size = 20),
          plot.title = element_text(size = 20), 
          axis.text.x = element_text(angle = 90, hjust = 1, size = 20),  
          axis.text.y = element_text(size = 20), 
          axis.title.x = element_text(size = 20), 
          axis.title.y = element_text(size = 20)) + 
  
          #facet_grid(. ~ clustering) + 
          ggtitle(taxa)
  
  
}
    
plots[[i]] <- plot
    
    i <- i + 1
    
}
```

```{r, echo=FALSE, warning=FALSE, message=FALSE, fig.width=20, fig.height=6}
for (i in 1:length(plots)) {
  plot(plots[[i]])
  cat("\r\n\r\n")
}
```