|paper|Cultivation and sequencing of rumen microbiome members from the Hungate1000 Collection|29-10-2018|
|---|---|---|

## Summary

* Collection of genomes from cultured organsisms, mostly from the rumen.

## Notes

* At the beginning of the project reference genomes were available for only 14 bacteria and one methanogen
* Members include 9 pyhla, 48 families and 82 genera
* Polysaccharide degrading genera were stated to be Butyrivibrio, Prevotella and Ruminococus
* Members of the Fermicutes and Bacteroidetes phyla predominate the rumen
* A core microbiome of 7 abundant genus-level groups was defined for 67% of the Global Rumen Census Sequences (in order):
  1. Prevotella (Bacteroidetes)
  2. Unclassified Clostridiales (Fermicutes)
  3. Unclassified Bacteroidetes
  4. Unclassified Ruminococcaceae (Fermicutes)
  5. Unclassified Lachnospiraceae (Fermicutes)
  6. Ruminococcus (Fermicutes)
  7. Butyrivibrio (Fermicutes)
* As an example, Bacteroides ovatus (over 320 glycoside hydrolases and lyases), where Lachnospiraceae bacterium has 296.
* The bacteroidetes have evolved polysaccharide utilization loci, genomis regions that encode all the required components for binding, transport, and depolymerization.
* Many hungate strains were shared between rumen and human instestinal microbiome samples.
* 47 isolates capable of B12 synthesis
* A further 140 genomes encode the enzymes for the salvage of B12 from an intermediate
* B12 required as a co-factor for enzymes involved in glucoegenesis from propionate in the liver.
* Rumen microbes are likely to be vertically inherited, whereas human intestinal isolates are transmitted from other sources in the environment.

[paper link] (https://www.nature.com/articles/nbt.4110)
